src/doc/tarpl/working-with-unsafe.md

```diff
@@ -0,0 +1,104 @@
+% Working with Unsafe
+
+Rust generally only gives us the tools to talk about Unsafe in a scoped and
+binary manner. Unfortunately, reality is significantly more complicated than that.
+For instance, consider the following toy function:
+
+```rust
+pub fn index(idx: usize, arr: &[u8]) -> Option<u8> {
+    if idx < arr.len() {
+        unsafe {
+            Some(*arr.get_unchecked(idx))
+        }
+    } else {
+        None
+    }
+}
+```
+
+Clearly, this function is safe. We check that the index is in bounds, and if it
+is, index into the array in an unchecked manner. But even in such a trivial
+function, the scope of the unsafe block is questionable. Consider changing the
+`<` to a `<=`:
+
+```rust
+pub fn index(idx: usize, arr: &[u8]) -> Option<u8> {
+    if idx <= arr.len() {
+        unsafe {
+            Some(*arr.get_unchecked(idx))
+        }
+    } else {
+        None
+    }
+}
+```
+
+This program is now unsound, and yet *we only modified safe code*. This is the
+fundamental problem of safety: it's non-local. The soundness of our unsafe
+operations necessarily depends on the state established by "safe" operations.
+Although safety *is* modular (we *still* don't need to worry about about
+unrelated safety issues like uninitialized memory), it quickly contaminates the
+surrounding code.
```

Comment:
Perhaps suggest that `unsafe` blocks should whenever possible enclose all code that can cause unsafety if poked at, specifically, all sanity checks before doing an unsafe operation should generally be kept in the same unsafe block. (As a style guideline)